x-pack/platform/plugins/shared/alerting/server/application/gap_auto_fill_scheduler/methods/update/update_gap_auto_fill_scheduler.ts

```diff
@@ -0,0 +1,217 @@
+  // Throw error if a rule type is not registered
+  for (const ruleTypeId of uniqueRuleTypeIds) {
+    context.ruleTypeRegistry.get(ruleTypeId);
+  }
+
+  const updatedRuleTypePairs = Array.from(new Set(updatedRuleTypes.map(toRuleTypeKey)));
+  if (updatedRuleTypePairs.length > 0) {
+    const filter = `(${updatedRuleTypePairs
+      .map(
+        (pair) =>
+          `${GAP_AUTO_FILL_SCHEDULER_SAVED_OBJECT_TYPE}.attributes.ruleTypeConsumerPairs: "${pair}"`
+      )
+      .join(' or ')})`;
+
+    const { saved_objects: candidates } = await soClient.find<GapAutoFillSchedulerSO>({
+      type: GAP_AUTO_FILL_SCHEDULER_SAVED_OBJECT_TYPE,
+      perPage: updatedRuleTypePairs.length + 1,
+      filter,
+    });
+
+    const hasConflictingPair = candidates.some((candidate) => candidate.id !== params.id);
+    if (hasConflictingPair) {
+      throw Boom.conflict(
+        `A gap auto fill scheduler already exists for at least one of the specified (rule type, consumer) pairs`
+      );
+    }
+  }
+
+  try {
+    const ruleTypesToCheck = buildRuleTypeUnion(existingRuleTypes, params.ruleTypes);
+    for (const ruleType of ruleTypesToCheck) {
+      await context.authorization.ensureAuthorized({
+        ruleTypeId: ruleType.type,
+        consumer: ruleType.consumer,
+        operation: WriteOperations.UpdateGapAutoFillScheduler,
+        entity: AlertingAuthorizationEntity.Rule,
+      });
+    }
+  } catch (error) {
+    context.auditLogger?.log(
+      gapAutoFillSchedulerAuditEvent({
+        action: GapAutoFillSchedulerAuditAction.UPDATE,
+        savedObject: {
+          type: GAP_AUTO_FILL_SCHEDULER_SAVED_OBJECT_TYPE,
+          id: params.id,
+          name: schedulerName,
+        },
+        error: error as Error,
+      })
+    );
+    throw error;
+  }
+  const now = new Date().toISOString();
+  const updatedEnabled = params.enabled;
+  const updatedSchedule = params.schedule;
+  const updatedScope = params.scope !== undefined ? params.scope : schedulerSo.attributes.scope;
+  const updatedAttributes: Partial<GapAutoFillSchedulerSO> = {
+    name: params.name,
+    enabled: updatedEnabled,
+    schedule: updatedSchedule,
+    gapFillRange: params.gapFillRange,
+    maxBackfills: params.maxBackfills,
+    numRetries: params.numRetries,
+    scope: updatedScope,
+    ruleTypes: updatedRuleTypes,
+    ruleTypeConsumerPairs: Array.from(new Set(updatedRuleTypes.map(toRuleTypeKey))),
+    updatedAt: now,
+    updatedBy: username ?? undefined,
+  };
+
+  try {
+    await soClient.update<GapAutoFillSchedulerSO>(
+      GAP_AUTO_FILL_SCHEDULER_SAVED_OBJECT_TYPE,
+      params.id,
+      updatedAttributes
+    );
+    const updatedSo = await soClient.get<GapAutoFillSchedulerSO>(
+      GAP_AUTO_FILL_SCHEDULER_SAVED_OBJECT_TYPE,
+      params.id
+    );
+    const nextSchedulerName = updatedSo.attributes.name ?? schedulerName;
+
+    await taskManager.removeIfExists(updatedSo.id);
+
+    if (updatedEnabled) {
+      await taskManager.ensureScheduled(
```

Comment:
We will be working on a bug fix to update TM tasks with API keys: https://github.com/elastic/kibana/issues/244918. When that is done, you should just be able to update the schedule of the existing task and not have to reschedule it.